File: _includes/profiles.pl
#!/usr/bin/perl
use strict;
use warnings;
use v5.10;
use Readonly;
use YAML qw{ LoadFile };
use Data::Dumper;
use URI::URL;
use LWP::Simple;

Readonly my $checkout      => "$ENV{HOME}/github/pdurbin/pdurbin.github.com";
Readonly my $googfav       => 'http://www.google.com/s2/u/0/favicons?domain=';
Readonly my $grabicon      => 'http://grabicon.com/';
Readonly my $favicondir    => 'images/favicons';
Readonly my $includedir    => "$checkout/_includes";
Readonly my $profileout    => "$includedir/profiles.html";
Readonly my $profiles_yaml => "$includedir/profiles.yaml";

chdir $checkout or die "Couldn't cd to $checkout";

mkdir $includedir unless -d $includedir;

my $profs = LoadFile($profiles_yaml);

my $html;

$html .= '<ul style="padding: 0px; margin: 0em; margin-left: 20px">' . "\n";
for my $proff (@$profs) {

    my $url = URI::URL->new($proff);

    my $googhost      = $googfav . $url->host;
    my $grabhost      = $grabicon . $url->host . '.png.16';
    my $url_host_real = $url->host;
    my $url_host_und  = $url_host_real;
    $url_host_und =~ s/\./_/;

    my $png = "$favicondir/$url_host_real.png";

    mkdir $favicondir unless -d $favicondir;

    unless ( -f $png ) {
        #system( "curl -s $googhost > $png");
        system("curl -s $grabhost > $png");
        sleep .5;
    }
    my $nohttp = $url->host . $url->path;

    my $li = qq{<li style="list-style-image:url('/$png')"><a href="$proff">$nohttp</a></li>};
    $html .= "$li\n";

}

$html .= "</ul>\n";

open( my $fh, '>', $profileout ) or die "cannot open > $profileout: $!";
print $fh $html;

File: _includes/profiles.py
#!/usr/bin/env python3
import os
import time
from pathlib import Path
from urllib.parse import urlparse

import requests
import yaml

HOME = os.environ.get("HOME", "")
checkout = Path(HOME) / "github" / "pdurbin" / "pdurbin.github.com"
googfav = "http://www.google.com/s2/u/0/favicons?domain="
grabicon = "http://grabicon.com/"
favicondir = Path("images") / "favicons"
includedir = checkout / "_includes"
profileout = includedir / "profiles.html"
profiles_yaml = includedir / "profiles.yaml"

checkout.mkdir(parents=True, exist_ok=True)
includedir.mkdir(parents=True, exist_ok=True)
favicondir.mkdir(parents=True, exist_ok=True)

os.chdir(checkout)

with profiles_yaml.open("r") as f:
    profs = yaml.safe_load(f)

html_lines = ['<ul style="padding: 0px; margin: 0em; margin-left: 20px">']

for proff in profs:
    parsed = urlparse(proff)
    host = parsed.hostname or parsed.netloc
    if not host:
        continue

    grabhost = f"{grabicon}{host}.png.16"

    png_path = favicondir / f"{host}.png"
    if not png_path.exists():
        try:
            resp = requests.get(grabhost, timeout=10)
            resp.raise_for_status()
            png_path.parent.mkdir(parents=True, exist_ok=True)
            with png_path.open("wb") as fh:
                fh.write(resp.content)
            time.sleep(0.5)
        except Exception:
            pass

    path = parsed.path
    if path == "/":
        path = ""
    nohttp = f"{host}{path}"

    png_url = "/" + png_path.as_posix()
    li = f'<li style="list-style-image:url(\'{png_url}\')"><a rel="me" href="{proff}">{nohttp}</a></li>'
    html_lines.append(li)

html_lines.append("</ul>")

with profileout.open("w") as fh:
    fh.write("\n".join(html_lines) + "\n")
